File: Hard/Java/First_Missing_Positive(41).java
class Solution {
    public int firstMissingPositive(int[] nums) {
        for (int i=0; i<nums.length; i++) {
            nums[i] = nums[i] <= 0 ? nums.length + 10 : nums[i];
        }
        for (int i=0; i<nums.length; i++) {
            if (Math.abs(nums[i]) <= nums.length) {
                nums[Math.abs(nums[i])-1] = -1*Math.abs(nums[Math.abs(nums[i])-1]);
            }
        }
        for (int i=0; i<nums.length; i++) {
            if (nums[i] > 0) {
                return i+1;
            }
        }
        return nums.length + 1;
    }
}

File: Hard/C++/First_Missing_Positive(41).cpp
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
	//removing negative numbers
	for (auto it = nums.begin(); it != nums.end(); )
		*it<=0?nums.erase(it) : it++;
    
	int n = nums.size(),x=0,b;

	//main code for finding missing number
	for(int i=0;;i++){
		b=nums[x];
		if(x==n)
			return n+1;
		if(nums[x]==x+1)
			x++;
		else if(nums[x]<=n && nums[x]!=nums[b-1])
			swap(nums[x],nums[nums[x]-1]);
		else
			(nums[n-1]==n)?swap(nums[x],nums[(n=n-2)]):swap(nums[x],nums[(n=n-1)]);
	}
}
};